cmd/k8s-operator/proxy.go

```diff
@@ -201,20 +168,128 @@ func runAPIServerProxy(s *tsnet.Server, rt http.RoundTripper, log *zap.SugaredLo
 		TLSNextProto: make(map[string]func(*http.Server, *tls.Conn, http.Handler)),
 		Handler:      ap,
 	}
-	log.Infof("listening on %s", ln.Addr())
+	log.Infof("API server proxy in %q mode is listening on %s", mode, ln.Addr())
 	if err := hs.ServeTLS(ln, "", ""); err != nil {
 		log.Fatalf("runAPIServerProxy: failed to serve %v", err)
 	}
 }
 
-const (
-	// oldCapabilityName is a legacy form of
-	// tailfcg.PeerCapabilityKubernetes capability. The only capability rule
-	// that is respected for this form is group impersonation - for
-	// backwards compatibility reasons.
-	// TODO (irbekrm): determine if anyone uses this and remove if possible.
-	oldCapabilityName = "https://" + tailcfg.PeerCapabilityKubernetes
-)
+// apiserverProxy is an http.Handler that authenticates requests using the Tailscale
+// LocalAPI and then proxies them to the Kubernetes API.
+type apiserverProxy struct {
+	log *zap.SugaredLogger
+	lc  *tailscale.LocalClient
+	rp  *httputil.ReverseProxy
+
+	mode        apiServerProxyMode
+	s           *tsnet.Server
+	upstreamURL *url.URL
+}
+
+func (h *apiserverProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	who, err := h.lc.WhoIs(r.Context(), r.RemoteAddr)
+	if err != nil {
+		h.log.Errorf("failed to authenticate caller: %v", err)
+		http.Error(w, "failed to authenticate caller", http.StatusInternalServerError)
+		return
+	}
+	if who == nil {
```

Comment:
In what cases might we expect both `err` and `who` to be nil? It seems kind of surprising to continue in that case. In `addImpersonationHeaders` we dereference `who` without checking for nil, but I'm not sure if that's a bug or not.

Reply:
Thanks for catching this - looks like [this errors if peer not found](https://github.com/tailscale/tailscale/pull/12466/commits/629465b8a5daefd148d6d5c8bf5bf6513836f129#diff-a1d5838382a0bbc737f1e7412428708251cb354429978538adba58f832d56deaR304), so I've removed the nil check.